consul/leader.go

```diff
@@ -562,11 +562,36 @@ func (s *Server) joinConsulServer(m serf.Member, parts *agent.Server) error {
 	}
 
 	// Attempt to add as a peer
-	addFuture := s.raft.AddPeer(raft.ServerAddress(addr))
-	if err := addFuture.Error(); err != nil {
-		s.logger.Printf("[ERR] consul: failed to add raft peer: %v", err)
+	minRaftProtocol, err := ServerMinRaftProtocol(s.serfLAN.Members())
+	if err != nil {
 		return err
 	}
+
+	if minRaftProtocol >= 2 && parts.RaftVersion >= 3 {
+		addFuture := s.raft.AddVoter(raft.ServerID(parts.ID), raft.ServerAddress(addr), 0, 0)
+		if err := addFuture.Error(); err != nil {
+			s.logger.Printf("[ERR] consul: failed to add raft peer: %v", err)
+			return err
+		}
+	} else {
+		addFuture := s.raft.AddPeer(raft.ServerAddress(addr))
```

Comment:
An interesting consequence here is that if they try to add an old server on v1, from a leader running v3 they will hit this case and get an unsupported protocol error, which is right.